src/Analyser/TypeSpecifierContext.php

```diff
@@ -61,7 +72,14 @@ public function negate(): self
 		if ($this->value === null) {
 			throw new ShouldNotHappenException();
 		}
-		return self::create(~$this->value & self::CONTEXT_BITMASK);
+
+		$negatedReturnType = null;
+		if ($this->returnType !== null) {
+			$baseType = $this->returnType->generalize(GeneralizePrecision::lessSpecific());
+			$negatedReturnType = TypeCombinator::remove($baseType, $this->returnType);
```

Comment:
This won't work for unions of int. We will need 2 two return types in the context and toggle between them on negate()